File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B10_analisis_factorial/python/B10_analisis_factorial.py
"""
B10. Análisis Factorial
Tema B10_analisis_factorial — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B10_analisis_factorial.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B10. Análisis Factorial")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B10. Análisis Factorial ═══
from scipy.optimize import minimize
from scipy.stats import chi2 as chi2_d

def factanal(R, m):
    """Análisis factorial por máxima verosimilitud (= factanal de R)."""
    p = R.shape[0]
    def crit(psi):
        psi = np.clip(psi, 1e-6, 1.0); sc = 1/np.sqrt(psi)
        ev = np.linalg.eigvalsh(sc[:,None]*R*sc[None,:])[::-1]
        e = ev[m:]; return -np.sum(np.log(e)-e) - (p-m)
    r = minimize(crit, np.diag(np.linalg.inv(R))**-1*(1-m/(2*p)),
                 method="L-BFGS-B", bounds=[(.005,1)]*p)
    psi = np.clip(r.x,.005,1); sc = 1/np.sqrt(psi)
    val, vec = np.linalg.eigh(sc[:,None]*R*sc[None,:])
    o = np.argsort(val)[::-1]; val, vec = val[o], vec[:,o]
    L = np.sqrt(psi)[:,None]*vec[:,:m]*np.sqrt(np.maximum(val[:m]-1,0))
    for j in range(m):                       # signo arbitrario: se fija por legibilidad
        if L[np.argmax(np.abs(L[:,j])),j] < 0: L[:,j] *= -1
    return L, psi, r.fun

def varimax(L, tol=1e-6, itmax=500):
    """Rotación varimax: empuja cada carga hacia 0 o ±1 para poder interpretar."""
    p_, k = L.shape; Rot = np.eye(k); d = 0.
    for _ in range(itmax):
        Lr = L@Rot
        u,s,vt = np.linalg.svd(L.T@(Lr**3 - Lr@np.diag(np.sum(Lr**2,0))/p_))
        Rot = u@vt; dn = np.sum(s)
        if d != 0 and dn/d < 1+tol: break
        d = dn
    return L@Rot

R_mat = np.corrcoef(X, rowvar=False)
n_obs, p_var = X.shape

# ── Verificación previa ────────────────────────────────────────────────
stat_b = -(n_obs-1-(2*p_var+5)/6)*np.linalg.slogdet(R_mat)[1]
p_b = 1-chi2_d.cdf(stat_b, p_var*(p_var-1)/2)
Ri = np.linalg.inv(R_mat); dd = np.sqrt(np.diag(Ri))
par = -Ri/np.outer(dd,dd); np.fill_diagonal(par,0)
Rc = R_mat.copy(); np.fill_diagonal(Rc,0)
kmo = np.sum(Rc**2)/(np.sum(Rc**2)+np.sum(par**2))

print(f"Bartlett: p = {p_b:.2e}  -> {'hay correlaciones que factorizar' if p_b<.05 else 'NADA que factorizar'}")
print(f"KMO     : {kmo:.3f}  (>0.6 aceptable)")

m_max = int(np.floor((2*p_var+1-np.sqrt(8*p_var+1))/2 - 1e-9))
print(f"Máximo de factores identificables: {m_max}  (con p={p_var} variables)")

# ── Prueba chi2 para elegir m ──────────────────────────────────────────
print(f"\n{'m':>2} | {'chi2':>9} | {'gl':>3} | {'p-valor':>9} | decisión")
print("-"*50)
for m in range(1, min(m_max, 5)+1):
    L,psi,f = factanal(R_mat, m)
    stat = (n_obs-1-(2*p_var+5)/6-2*m/3)*f
    gl = int(((p_var-m)**2-p_var-m)/2)
    pv = 1-chi2_d.cdf(stat, gl) if gl>0 else np.nan
    print(f"{m:2d} | {stat:9.2f} | {gl:3d} | {pv:9.4f} | {'ADECUADO' if pv>=.05 else 'insuficiente'}")
print("⚠️  p-valor ALTO = el modelo SÍ es adecuado (al revés de lo habitual)")

# ── Ajuste final con rotación varimax ──────────────────────────────────
M = min(2, m_max)                      # ⬅ elige m según la tabla de arriba
L, psi, _ = factanal(R_mat, M)
Lv = varimax(L) if M > 1 else L

tab = pd.DataFrame(Lv, index=nombres_X, columns=[f"Factor{i+1}" for i in range(M)])
tab["comunalidad"] = (Lv**2).sum(1)    # varianza COMPARTIDA
tab["unicidad"] = psi                  # varianza PROPIA (PCA no separa esto)
print(f"\nCargas con rotación varimax (m={M}):")
print(tab.round(3).to_string())
print("\ncomunalidad + unicidad = 1. Cargas |·|>0.3 son las que 'definen' el factor.")

fig, ax = plt.subplots(figsize=(6, max(3, .35*len(nombres_X))))
im = ax.imshow(Lv, cmap="RdBu_r", vmin=-1, vmax=1, aspect="auto")
ax.set_xticks(range(M)); ax.set_xticklabels([f"F{i+1}" for i in range(M)])
ax.set_yticks(range(len(nombres_X))); ax.set_yticklabels(nombres_X, fontsize=8)
for i in range(len(nombres_X)):
    for j in range(M):
        ax.text(j,i,f"{Lv[i,j]:.2f}",ha="center",va="center",fontsize=8,
                color="white" if abs(Lv[i,j])>.5 else "black")
plt.colorbar(im, label="carga"); ax.set_title("Cargas factoriales (varimax)")
plt.tight_layout(); plt.show()

print("\nListo. Revisa la carpeta figuras/")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B10_analisis_factorial/R/B10_analisis_factorial.R
# ############################################################################
# B10. ANÁLISIS FACTORIAL
# Tema B10_analisis_factorial — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B10_analisis_factorial.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B10. ANÁLISIS FACTORIAL\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B10. ANÁLISIS FACTORIAL
# ----------------------------------------------------------------------------
# QUÉ HACE: PCA resume variables. El análisis factorial va más allá: supone
# que hay CAUSAS OCULTAS (factores latentes) que PRODUCEN lo que observas.
# Ejemplo clásico: 9 pruebas de habilidad mental; detrás hay 3 factores no
# observables (verbal, espacial, numérico) que causan los puntajes.
#
#   X = L F + e        con    Sigma = L L' + Psi
#     L   = CARGAS: cuánto pesa cada factor sobre cada variable
#     Psi = UNICIDADES: la parte de cada variable que no comparte con nadie
#
# ⚠️ LA DIFERENCIA CON PCA (lo preguntan)
#              PCA                     ANÁLISIS FACTORIAL
#   Dirección  componentes = CONSECUENCIA   factores = CAUSA
#   Varianza   modela TODA                  solo la COMPARTIDA (separa ruido)
#   Solución   única y cerrada              iterativa, NO ÚNICA
#   Objetivo   comprimir                    INTERPRETAR
#
# LA ROTACIÓN: la solución no es única (si L sirve, L*R también). Se
# aprovecha para hacer los factores interpretables.
#   "varimax" (ortogonal): factores NO correlacionados. El default.
#   "promax"  (oblicua)  : permite factores correlacionados.
#   LA ROTACIÓN NO CAMBIA EL AJUSTE, solo el sistema de ejes.
#
# ⚠️ LA PRUEBA CHI-CUADRADO
#   H0: m factores bastan.
#   P-VALOR ALTO = EL MODELO CON m FACTORES SÍ ES ADECUADO.
#   Es AL REVÉS de la lectura habitual de un p-valor.
#
# ⚠️ LÍMITE DE IDENTIFICABILIDAD: m < (2p + 1 - sqrt(8p+1))/2
#   Con p=4 el máximo es 1 factor. Por eso factanal(iris,2) da ERROR:
#   NO es un bug, es una restricción del modelo.
#
# VERIFICACIÓN PREVIA
#   Bartlett: H0 = no hay nada que factorizar. Se quiere RECHAZAR.
#   KMO: >0.8 excelente · >0.6 aceptable · <0.5 inadecuado
#
# PARÁMETROS de factanal()
#   factors  : número de factores m
#   rotation : "varimax" | "promax" | "none"
#   scores   : "regression" o "Bartlett" para obtener los scores
#   covmat + n.obs : permite ajustar SOLO con la matriz de correlación
# ============================================================================

library(psych)

p_var <- ncol(X); n_obs <- nrow(X)
m_max <- floor((2*p_var + 1 - sqrt(8*p_var + 1))/2 - 1e-9)
cat(sprintf("Máximo de factores identificables: %d (con p=%d variables)\n", m_max, p_var))

# Verificación previa
print(cortest.bartlett(cor(X), n=n_obs))    # se quiere p-valor PEQUEÑO
cat(sprintf("KMO: %.3f\n", KMO(cor(X))$MSA))

# Prueba chi2 para elegir m
cat("\nDeterminando el número de factores:\n")
for (m in 1:min(m_max, 5)) {
  ff <- try(factanal(covmat=cor(X), factors=m, n.obs=n_obs), silent=TRUE)
  if (!inherits(ff, "try-error"))
    cat(sprintf("  m=%d  chi2=%8.2f  gl=%3d  p=%.4f  %s\n", m, ff$STATISTIC,
                ff$dof, ff$PVAL, ifelse(ff$PVAL>=.05,"ADECUADO","insuficiente")))
}
cat("⚠️ p-valor ALTO = el modelo SÍ es adecuado (al revés de lo habitual)\n")

# Ajuste final con rotación
M <- min(2, m_max)                          # ⬅ elige m según la tabla
fa <- factanal(X, factors=M, rotation="varimax", scores="regression")
print(fa)
cat("\ncomunalidad + unicidad = 1. Cargas |·|>0.3 definen el factor.\n")

# Compara con rotación promax (oblicua)
# fa2 <- factanal(X, factors=M, rotation="promax"); print(fa2)

cat("\nListo. Revisa la carpeta figuras/\n")
